File: cpp/Stack/implement_queue_using_stacks.cpp
#include <deque>

class Queue {
private:
    std::deque<int> dq;
public:
    // Push element x to the back of queue.
    void push(int x) {
        dq.push_back(x);
    }

    // Removes the element from in front of queue.
    void pop(void) {
        dq.pop_front();
    }

    // Get the front element.
    int peek(void) {
        return dq.front();
    }

    // Return whether the queue is empty.
    bool empty(void) {
        return dq.empty();
    }
};

File: py/stack/ImplementQueueusingStacks.py
class MyQueue(object):

    def __init__(self):
        self.activeBlock = []
        self.inactiveBlock = []
        self.peekVal = 0
        
    def push(self, x):
        if len(self.activeBlock) == 0:
            self.peekVal = x
        self.activeBlock.append(x)
        
    def pop(self):
        if len(self.activeBlock) > 0:
            while len(self.activeBlock) > 0:
                self.inactiveBlock.append(self.activeBlock.pop())
            x = self.inactiveBlock.pop()
            if len(self.inactiveBlock) > 0:
                self.peekVal = self.inactiveBlock.pop()
                self.activeBlock.append(self.peekVal)
            while len(self.inactiveBlock) > 0:
                self.activeBlock.append(self.inactiveBlock.pop())
            return x
        
    def peek(self):
        if len(self.activeBlock) > 0:
            return self.peekVal
        
    def empty(self):
        return len(self.activeBlock) == 0
